File: src/Dorc.Terraform.Catalog/TerraformTemplateManifest.cs
namespace Dorc.Terraform.Catalog
{
    // Manifest for one published version of a stock template.
    // Resolved by ITemplateCatalog implementations (e.g. GitTemplateCatalog).
    // Schema mirrored as docs/Terraform/MANIFEST-SCHEMA.json (planned).
    // additive fields: TerraformTemplateSource.SubPath defaults to null 
    // the resolver falls back to stock-modules/<Name>. TerraformTemplateParameter.Sensitive
    // defaults to false.
    public sealed record TerraformTemplateManifest(
        string Name,
        string Version,
        TerraformTemplateSource Source,
        IReadOnlyList<TerraformTemplateParameter> Parameters,
        IReadOnlyList<TerraformTemplateOutput> Outputs,
        string? Description,
        IReadOnlyList<string> Tags,
        string? Category,
        IReadOnlyDictionary<string, string> RequiredProviders,
        string RequiredTerraformVersion,
        string? Owner,
        bool Deprecated,
        string? DeprecationReason);

    public sealed record TerraformTemplateSource(
        string Kind,
        string Locator,
        string Ref,
        string? SubPath = null);

    public sealed record TerraformTemplateParameter(
        string Name,
        TerraformParameterType Type,
        bool Required,
        string? Description,
        string? Default,
        IReadOnlyList<string>? AllowedValues,
        string? Pattern,
        decimal? Min,
        decimal? Max,
        bool Sensitive = false);

    public sealed record TerraformTemplateOutput(
        string Name,
        TerraformParameterType Type,
        string? Description,
        bool Sensitive);

    public enum TerraformParameterType
    {
        String,
        Number,
        Bool,
        List,
        Map,
        Object,
    }
}

File: src/dorc-web/src/apis/dorc-api/models/TerraformTemplateManifest.ts
// Hand-written extension matching Dorc.Terraform.Catalog.TerraformTemplateManifest.
// When the openapi codegen runs, this file should be regenerated and the hand-
// written shape kept in sync with the C# record.

export interface TerraformTemplateSource {
    Kind: string;
    Locator: string;
    Ref: string;
    SubPath?: string | null;
}

export type TerraformParameterType = 'String' | 'Number' | 'Bool' | 'List' | 'Map' | 'Object';

export interface TerraformTemplateParameter {
    Name: string;
    Type: TerraformParameterType;
    Required: boolean;
    Description?: string | null;
    Default?: string | null;
    AllowedValues?: string[] | null;
    Pattern?: string | null;
    Min?: number | null;
    Max?: number | null;
    Sensitive: boolean;
}

export interface TerraformTemplateOutput {
    Name: string;
    Type: TerraformParameterType;
    Description?: string | null;
    Sensitive: boolean;
}

export interface TerraformTemplateManifest {
    Name: string;
    Version: string;
    Source: TerraformTemplateSource;
    Parameters: TerraformTemplateParameter[];
    Outputs: TerraformTemplateOutput[];
    Description?: string | null;
    Tags: string[];
    Category?: string | null;
    RequiredProviders: { [key: string]: string };
    RequiredTerraformVersion: string;
    Owner?: string | null;
    Deprecated: boolean;
    DeprecationReason?: string | null;
}
